File: apps/web/src/lib/core/text_limits.test.ts
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { readFileSync, readdirSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { join } from 'node:path';
import { TEXT_LIMITS, TEXT_LIMIT_CONSTRAINTS } from './text_limits.js';

// Relative to this file's own directory (`apps/web/src/lib/core/`), so it
// resolves the same however the suite is invoked: core → lib → src → web → apps.
const MIGRATIONS = fileURLToPath(
	new URL('../../../../backend/supabase/migrations/', import.meta.url)
);

function migrationsSql(): string {
	return readdirSync(MIGRATIONS)
		.filter((f) => f.endsWith('.sql'))
		.sort()
		.map((f) => readFileSync(join(MIGRATIONS, f), 'utf8'))
		.join('\n');
}

/**
 * The caps in `text_limits.ts` are only useful if they are the SAME numbers the
 * database enforces. A composer capped above the constraint hands the user a
 * 23514 they cannot act on; one capped below silently truncates. So the guard
 * reads the cap out of the migrations rather than restating it.
 *
 * The whole directory, not one file: the registered caps are added by three
 * different migrations, and a guard pointed at one of them can only ever
 * certify the caps that migration happened to contain.
 */
function capsFromMigrations(): Record<string, number> {
	const caps: Record<string, number> = {};
	// `add constraint <name> check (… char_length(<col>) <= <n>) not valid;`
	const re =
		/add\s+constraint\s+(\w+)\s+check\s*\([^;]*?char_length\([^)]*\)\s*<=\s*(\d+)/gi;
	for (const match of migrationsSql().matchAll(re)) caps[match[1]] = Number(match[2]);
	return caps;
}

test('parsed a cap out of the migrations for every registered constraint', () => {
	// Without this the per-key loop below would pass over nothing if the regex
	// ever stopped matching (decisions § 534).
	const caps = capsFromMigrations();
	for (const constraint of Object.values(TEXT_LIMIT_CONSTRAINTS)) {
		assert.ok(constraint in caps, `no CHECK found for ${constraint}`);
	}
});

test('every client cap equals the constraint the database enforces', () => {
	const caps = capsFromMigrations();
	for (const [key, constraint] of Object.entries(TEXT_LIMIT_CONSTRAINTS)) {
		assert.equal(
			caps[constraint],
			TEXT_LIMITS[key as keyof typeof TEXT_LIMITS],
			`${key} vs ${constraint}`
		);
	}
});

test('the migrations emit a VALIDATE for every registered constraint', () => {
	const sql = migrationsSql();
	// `20261124_001` added three NOT VALID caps and never validated any of them,
	// so those rows stayed permanently unchecked until `20270503_001` caught up.
	// A cap this registry claims must be one the database actually enforces on
	// the rows already there.
	for (const constraint of Object.values(TEXT_LIMIT_CONSTRAINTS)) {
		assert.ok(
			sql.includes(`validate constraint ${constraint}`),
			`no VALIDATE for ${constraint}`
		);
	}
});

File: apps/mobile_android/test/text_limits_test.dart
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';

import '../lib/text_limits.dart';

/// The caps in `text_limits.dart` are only useful if they are the SAME numbers
/// the database enforces, so the guard reads them out of the migrations rather
/// than restating them. The whole directory, not one file: the registered caps
/// are added by three different migrations, and a guard pointed at one of them
/// can only ever certify the caps that migration happened to contain.
/// Mirrors `apps/web/src/lib/core/text_limits.test.ts`.
const _migrationsDir = 'apps/backend/supabase/migrations';

Directory _migrations() {
  // The suite runs from `apps/mobile_android` on Android and
  // `apps/mobile_ios` on the twin, so walk up to the repo root.
  var dir = Directory.current;
  for (var i = 0; i < 4; i++) {
    final d = Directory('${dir.path}/$_migrationsDir');
    if (d.existsSync()) return d;
    dir = dir.parent;
  }
  fail('could not locate $_migrationsDir from ${Directory.current.path}');
}

String _migrationsSql() {
  final files = _migrations()
      .listSync()
      .whereType<File>()
      .where((f) => f.path.endsWith('.sql'))
      .toList()
    ..sort((a, b) => a.path.compareTo(b.path));
  return files.map((f) => f.readAsStringSync()).join('\n');
}

Map<String, int> _capsFromMigrations() {
  final sql = _migrationsSql();
  final re = RegExp(
    r'add\s+constraint\s+(\w+)\s+check\s*\([^;]*?char_length\([^)]*\)\s*<=\s*(\d+)',
    caseSensitive: false,
  );
  return {
    for (final m in re.allMatches(sql))
      m.group(1)!: int.parse(m.group(2)!),
  };
}

void main() {
  group('text limits vs the database CHECK constraints', () {
    final caps = _capsFromMigrations();

    test('parsed a cap out of the migrations for every registered constraint',
        () {
      // Without this the per-key loop would pass over nothing if the pattern
      // ever stopped matching (decisions § 534).
      for (final constraint in kTextLimitConstraints.keys) {
        expect(caps, contains(constraint));
      }
    });

    kTextLimitConstraints.forEach((constraint, cap) {
      test('$constraint matches the client cap $cap', () {
        expect(caps[constraint], cap);
      });
    });

    test('the migrations emit a VALIDATE for every registered constraint', () {
      final sql = _migrationsSql();
      for (final constraint in kTextLimitConstraints.keys) {
        expect(sql, contains('validate constraint $constraint'));
      }
    });
  });
}
